pkg/agent/multicluster/pod_route_controller.go

```diff
@@ -0,0 +1,394 @@
+		klog.ErrorS(err, "Error syncing Gateway, requeuing", "key", key)
+	}
+	return true
+}
+
+func (c *MCPodRouteController) syncGateway() error {
+	activeGW, err := getActiveGateway(c.gwLister)
+	if err != nil {
+		klog.ErrorS(err, "Failed to get an active Gateway")
+		return err
+	}
+
+	c.podWorkersStartedMutex.Lock()
+	defer c.podWorkersStartedMutex.Unlock()
+
+	amIGateway := activeGW != nil && c.nodeConfig.Name == activeGW.Name
+	// Stop Pod flow controller and clean up all installed Multi-cluster Pod flows,
+	// if the Node was a Gateway before.
+	if !amIGateway {
+		if c.podWorkersStarted {
+			klog.InfoS("Shutting down Multi-cluster PodFlowController")
+			close(c.podWorkerStopCh)
+			c.podWorkerStopCh = nil
+			c.podInformer = nil
+			c.podLister = nil
+			c.podWorkersStarted = false
+		}
+	}
+
+	if !amIGateway || (amIGateway && !c.podWorkersStarted) {
+		err := c.ofClient.UninstallMulticlusterPodFlows("")
+		if err != nil {
+			return err
+		}
+	}
+
+	if amIGateway {
+		if !c.podWorkersStarted {
+			klog.InfoS("Starting Multi-cluster PodFlowController")
+			c.podWorkerStopCh = make(chan struct{})
+			c.createPodInformer()
+			go c.podInformer.Run(c.podWorkerStopCh)
+			if !cache.WaitForNamedCacheSync(podRouteControllerName, c.podWorkerStopCh, c.podInformer.HasSynced) {
+				c.podWorkerStopCh = nil
+				c.podInformer = nil
+				c.podLister = nil
+				return errors.New("failed to sync Pod cache")
+			}
+
+			for i := 0; i < podWorkerNum; i++ {
+				go wait.Until(c.podWorker, time.Second, c.podWorkerStopCh)
+			}
+			c.podWorkersStarted = true
+			return nil
+		}
+		// Do nothing when the Pod flow controller is already started since
+		// Pod flow controller will be responsible for handling Pod events to install flows.
+	}
+	return nil
+}
+
+func (c *MCPodRouteController) podWorker() {
+	for c.processPodNextWorkItem() {
+	}
+}
+
+func (c *MCPodRouteController) processPodNextWorkItem() bool {
+	obj, quit := c.podQueue.Get()
+	if quit {
+		return false
+	}
+	defer c.podQueue.Done(obj)
+
+	if k, ok := obj.(string); !ok {
+		c.podQueue.Forget(obj)
+		klog.InfoS("Expected string in work queue but got", "object", obj)
+		return true
+	} else if err := c.syncPod(k); err == nil {
+		c.podQueue.Forget(k)
+	} else {
+		c.podQueue.AddRateLimited(k)
+		klog.ErrorS(err, "Error syncing key, requeuing", "key", k)
+	}
+	return true
+}
+
+func (c *MCPodRouteController) syncPod(podIP string) error {
+	c.podWorkersStartedMutex.RLock()
+	defer c.podWorkersStartedMutex.RUnlock()
+	if !c.podWorkersStarted {
+		return nil
+	}
+
+	pods, _ := c.podInformer.GetIndexer().ByIndex(podIndexKey, podIP)
+	if len(pods) == 0 {
+		klog.V(2).InfoS("Deleting Multi-cluster flows for Pod", "podIP", podIP)
+		if err := c.ofClient.UninstallMulticlusterPodFlows(podIP); err != nil {
+			klog.ErrorS(err, "Failed to uninstall Multi-cluster flows for Pod", "podIP", podIP)
+			return err
+		}
+		return nil
+	}
+
+	latestPod := c.getLatestPod(pods)
+	nodeIP := latestPod.Status.HostIP
```

Comment:
It's possible HostIP is empty at this moment as Pod could be updated asynchronously. You could change `getlatestPod` to `getLatestValidPod` and move `c.podInformer.GetIndexer().ByIndex(podIndexKey, podIP)` to it too, so Uninstall case can be handled together.

Reply:
I suppose this won't happen since `podIPIndexFunc` will also validate Pod, the Pod won't be added to the cache if there is no PodIP or NodeIP or it's hostNetwork.

Reply:
You are right.